File: packages/guards/src/phone-country-check.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface PhoneCountryCheckOptions {
  action: 'block' | 'warn';
  allowedCountryCodes?: string[];
}

const PREMIUM_PREFIXES = ['900', '976', '809', '284', '876'];

const PHONE_RE = /\+(\d{1,3})[\s.-]?\(?\d{1,4}\)?[\s.-]?\d{2,4}[\s.-]?\d{2,8}/g;

export function phoneCountryCheck(options: PhoneCountryCheckOptions): Guard {
  return {
    name: 'phone-country-check',
    version: '0.1.0',
    description: 'Validates phone country codes and detects premium rate numbers',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      let match: RegExpExecArray | null;
      const re = new RegExp(PHONE_RE.source, PHONE_RE.flags);
      while ((match = re.exec(text)) !== null) {
        const countryCode = match[1];
        const fullNum = match[0].replace(/[\s.()+\-]/g, '');

        if (options.allowedCountryCodes?.length) {
          if (!options.allowedCountryCodes.includes(countryCode)) {
            issues.push(`restricted_country:+${countryCode}`);
          }
        }

        for (const prefix of PREMIUM_PREFIXES) {
          if (fullNum.startsWith(`1${prefix}`) || fullNum.startsWith(prefix)) {
            issues.push(`premium_rate:${prefix}`);
          }
        }
      }

      const triggered = issues.length > 0;
      return {
        guardName: 'phone-country-check',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min(issues.length / 3, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/phone_country_check.py
"""Validate phone country codes and detect premium rate numbers."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PREMIUM_PREFIXES = ["900", "976", "809", "284", "876"]

_PHONE_RE = re.compile(
    r"\+(\d{1,3})[\s.-]?"
    r"\(?\d{1,4}\)?[\s.-]?"
    r"\d{2,4}[\s.-]?\d{2,8}"
)


class _PhoneCountryCheck:
    def __init__(
        self,
        *,
        action: str = "block",
        allowed_country_codes: Optional[List[str]] = None,
    ) -> None:
        self.name = "phone-country-check"
        self.action = action
        self._allowed = allowed_country_codes or []

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        for m in _PHONE_RE.finditer(text):
            cc = m.group(1)
            full = re.sub(r"[\s.()+\-]", "", m.group(0))

            if self._allowed and cc not in self._allowed:
                issues.append(f"restricted_country:+{cc}")

            for prefix in _PREMIUM_PREFIXES:
                if (
                    full.startswith(f"1{prefix}")
                    or full.startswith(prefix)
                ):
                    issues.append(f"premium_rate:{prefix}")

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="phone-country-check",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=(
                min(len(issues) / 3, 1.0)
                if triggered
                else 0.0
            ),
            message=(
                "Phone country code issue detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def phone_country_check(
    *,
    action: str = "block",
    allowed_country_codes: Optional[List[str]] = None,
) -> _PhoneCountryCheck:
    return _PhoneCountryCheck(
        action=action,
        allowed_country_codes=allowed_country_codes,
    )
